---
title: "An Introduction to MMAPPR2"
author:
- name: Kyle Johnsen
  affiliation: Brigham Young University, Provo, UT
package: MMAPPR2
output: 
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document: default
abstract: |
  Instructions on mapping mutations from forward genetic screens using MMAPPR2.
vignette: |
  %\VignetteIndexEntry{An Introduction to MMAPPR2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

# Getting Started

## Installing Dependencies
You must first install Ensembl VEP like so:
```{bash installVEP}
git clone https://github.com/Ensembl/ensembl-vep.git
cd ensembl-vep
perl INSTALL.pl -a ac -s my_species
```

This installs the most recent VEP and cache for your desired species, which is what MMAPPR2 expects by default.
If you depart from the installation shown here, see [Ensembl's instructions](http://www.ensembl.org/info/docs/tools/vep/script/vep_download.html#installer)
and make sure any differences are accounted for in the
[`VEPFlags`](#configure-vepflags-object) object.

## Reference Genome
Variant calling requires a reference genome, which you can create from your reference file like this:
```{r refGenome eval=FALSE}
refGenome <- gmapR::GmapGenome('correct_genome.fa', create=TRUE)
```

**Make sure your reference genome is the same one installed by VEP, i.e., the most recent one on Ensembl!**

## Running MMAPPR2

Setting up MmapprParam
```{r param}
library(MMAPPR2)

MMAPPR2:::.insertFakeVEPintoPath() ## Don't worry about this line
param <- MmapprParam(refGenome = gmapR::GmapGenome('GRCz11'),
                     wtFiles = 'tests/testthat/test_data/bam_files/zy14_wt_cut_filt.bam',
                     mutFiles = 'tests/testthat/test_data/bam_files/zy14_mut_cut_filt.bam',
                     
```

Do stuff


## Advanced Configuration

#### Configure VEPFlags Object
MMAPPR2 uses the ensemblVEP Bioconductor package to predict the effect of variants in the peak region. 
To customize this process, you'll need to configure a `VEPFlags` object. Look at [Ensembl's website for script options](http://uswest.ensembl.org/info/docs/tools/vep/script/vep_options.html). You can configure the `VEPFlags` object like this:
```{r vepFlags}
library(ensemblVEP)
vepFlags <- VEPFlags(flags = list(
    ### DEFAULT SETTINGS
    format = 'vcf',  # <-- this is necessary
    vcf = FALSE,  # <-- as well as this
    species = 'danio_rerio',
    database = FALSE,
    cache = TRUE,
    filter_common = TRUE,
    coding_only = TRUE  # assuming RNA-seq data
    ### YOU MAY FIND THESE INTERESTING:
    # everything = TRUE  # enables many optional analyses, such as Polyphen and SIFT
    # per_gene = TRUE  # will output only the worst variant per gene
    # pick = TRUE  # will output only one consequence per variant
))
```

#### BiocParallel Configuration
MMAPPR2 simply uses the default `bpparam` registered. You can change this (for example, if BiocParallel isn't working correctly) with the `BiocParallel::register` command. For example:
```{r bpparam}
library(BiocParallel)
register(SerialParam())
register(MultiCoreParam())
```

The last param registered becomes the default.

#### Whole-genome Sequencing (WGS)
It would probably work. Mess with parameters like minDepth.
