---
title: "An Introduction to MMAPPR2"
author:
- name: Kyle Johnsen
  affiliation: Brigham Young University, Provo, UT
package: MMAPPR2
output: 
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document: default
abstract: |
  Instructions on mapping mutations from forward genetic screens using MMAPPR2.
vignette: |
  %\VignetteIndexEntry{An Introduction to MMAPPR2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

# Getting Started

## Installing Dependencies
You'll need Ensembl VEP, which you can install like this:
```{r installVEP, eval=FALSE}
git clone https://github.com/Ensembl/ensembl-vep.git
cd ensembl-vep
perl INSTALL.pl -a ac -s my_species
```

This installs the most recent VEP and cache for your desired species, which is what MMAPPR2 expects by default.
If you depart from the installation shown here, or if things don't go smoothly, see [Ensembl's instructions](http://www.ensembl.org/info/docs/tools/vep/script/vep_download.html#installer)
and make sure any differences are accounted for in the
[`VEPFlags`](#configure-vepflags-object) object.

## Reference Genome
Variant calling requires a reference genome, which you can create from your reference file like this:
```{r refGenome, echo=TRUE, results='hide'}
library(MMAPPR2data, quietly = TRUE)

refGenome <- gmapR::GmapGenome(goldenFasta(), name='slc24a5', create=TRUE)
```

For our example, we will use just the golden gene from the GRCz11 zebrafish reference genome. **Make sure your reference genome is the same you'll use with VEP! This will be the most recent assembly available on Ensembl unless you customize. You should use the same genome in aligning your sequencing data as well.**

# Basic Use

## Setting Parameters

Here we configure the VEPFlags object to use our example fasta and GTF files.
[See below for more info.](#configure-vepflags-object)

```{r param}
BiocParallel::register(BiocParallel::MulticoreParam())  ## see below for explanation of BiocParallel
library(MMAPPR2, quietly = TRUE)
library(Rsamtools, quietly = TRUE)

vepFlags <- ensemblVEP::VEPFlags(flags = list(
    format = 'vcf',  # <-- this is necessary
    vcf = FALSE,  # <-- as well as this
    database = FALSE,
    species = 'danio_rerio',
    fasta = goldenFasta(),
    gff = goldenGFF(),
    filter_common = TRUE,
    coding_only = TRUE  # assuming RNA-seq data
))

param <- MmapprParam(refGenome = gmapR::GmapGenome('slc24a5'),
                     wtFiles = exampleWTbam(),
                     mutFiles = exampleMutBam(),
                     species = 'danio_rerio',
                     vepFlags = vepFlags,
                     outputFolder = tempdir())
```

## Running MMAPPR2
With parameters set, running the pipeline should be as simple as the following:
```{r mmappr}
mmapprData <- mmappr(param)
```

The MMAPPR2 pipeline can also be run a step at a time:
```{r mmappr-steps}
md <- new('MmapprData', param=param) ## calculateDistance() takes a MmapprData object
postCalcDistMD <- calculateDistance(md)
postLoessMD <- loessFit(postCalcDistMD)
postPrePeakMD <- prePeak(postLoessMD)
postPeakRefMD <- peakRefinement(postPrePeakMD)
postCandidatesMD <- generateCandidates(postPeakRefMD)
outputMmapprData(postCandidatesMD)
```

If the pipeline fails midway, the `MmapprData` object is saved, which you can then load
and use for inspection and debugging:
```{r recover-md}
## Contents of output folder:
system2('ls', outputFolder(param(mmapprData)))

mdFile <- file.path(outputFolder(param(mmapprData)), 'mmappr_data.RDS')
md <- readRDS(mdFile)
md
```


# Advanced Configuration

## Configure VEPFlags Object
MMAPPR2 uses the `r BiocStyle::Biocpkg("ensemblVEP")` Bioconductor package to predict the effect of variants in the peak region. 
To customize this process, you'll need to configure a `VEPFlags` object. Look at [Ensembl's website for script options](http://uswest.ensembl.org/info/docs/tools/vep/script/vep_options.html). You can configure the `VEPFlags` object like this:
```{r vepFlags}
library(ensemblVEP, quietly = TRUE)
vepFlags <- VEPFlags(flags = list(
    ### DEFAULT SETTINGS
    format = 'vcf',  # <-- this is necessary
    vcf = FALSE,  # <-- as well as this
    species = 'danio_rerio',
    database = FALSE,
    cache = TRUE,
    filter_common = TRUE,
    coding_only = TRUE  # assuming RNA-seq data
    ### YOU MAY FIND THESE INTERESTING:
    # everything = TRUE  # enables many optional analyses, such as Polyphen and SIFT
    # per_gene = TRUE  # will output only the most severe variant per gene
    # pick = TRUE  # will output only one consequence per variant
))
```

## BiocParallel Configuration
MMAPPR2 simply uses the default `bpparam` registered. You can change this (for example, if `r BiocStyle::Biocpkg("BiocParallel")` isn't working correctly) with the `BiocParallel::register` command. For example:
```{r bpparam}
library(BiocParallel, quietly = TRUE)
register(SerialParam())
register(MulticoreParam(progressbar=TRUE))
registered()
```

The last param registered becomes the default.

## Whole-genome Sequencing (WGS)
MMAPPR2, like its predecessor, was designed for and tested using RNA-Seq data. However, the principles at work should still apply for WGS data.


-----------------
# Session Info
```{r sessionInfo}
sessionInfo()
```
